Proposal: Add autoconfiguration for Symfony 4, other tweaks

adiq opened the pull request:
So this is my proposal for SymfonyExtension. Related to #28 

What it changes:

1. Adds autoconfiguration for Symfony 4 (when default `src/Kernel.php` is detected); Improves DX.

So when project bein compliant with new Symfony 4 structure, the configuration will look like:
```
            extensions:
                FriendsOfBehat\SymfonyExtension: ~
```
The standard `.env` file is loaded by default.

2. Explicit settings of `env` and `debug` parameters will override the configuration loaded from `.env` file. This also improves DX, as there is no need for hacky workarounds to get behat working.

3. Fixed issue with the badly named configuration `$config['kernel']['kernel'] = $debugMode;` in 1a39d31f5e457cc6b44701a7bef99ba14f5c0cb2

4. Tweaked README to include new configuration details


When it might seem only like an improvement, it's required to get behat working on new Symfony 4 structure when you: do not want to store a separate copy of `.env` with only `APP_ENV` changed to `test`, do not want to change those settings back and forth in `.env` file; which I consider not good practice and not good DX for sure 😄

adiq commented:
@pamil could you spare a moment to look at this? I would prefer to use official package over my forked repo 😸

pamil commented:
Hello, I'm just back from holiday and trying to catch up with all the things, I'll try to process this PR this week :)

kochen commented on README.md:
```diff
@@ -26,30 +26,40 @@ ensures that application behaviour will not be affected by stateful services.
         path: src/Kernel.php
-        debug: true
+        class: 'App\Kernel'
+        env: test # When explicitly set, will overrite APP_ENV loaded from env_file file
```
spelling: `overwrite` or `override`
(but not `overrite`)

kochen commented on README.md:
```diff
@@ -26,30 +26,40 @@ ensures that application behaviour will not be affected by stateful services.
-        debug: true
+        class: 'App\Kernel'
+        env: test # When explicitly set, will overrite APP_ENV loaded from env_file file
+        debug: true  # When explicitly set, will overrite APP_DEBUG loaded from env_file file
```
spelling: `overwrite` or `override`
(but not `overrite`)

kochen commented on src/ServiceContainer/SymfonyExtension.php:
```diff
@@ -99,18 +127,15 @@ public function initialize(ExtensionManager $extensionManager): void
+                        ->scalarNode('bootstrap')->end()
+                        ->scalarNode('path')->end()
+                        ->scalarNode('class')->end()
+                        ->scalarNode('env')->end()
```
default to `test` if omitted

adiq replied:
This default is defined in `SYMFONY_DEFAULTS` / `SYMFONY_4_DEFAULTS` -> `DEFAULT_ENV`. We can't set a default on input loading because we are losing information if given value has been entered by the user or is a default one (and we need that information in `autoconfigure` method).

kochen commented on src/ServiceContainer/SymfonyExtension.php:
```diff
@@ -99,18 +127,15 @@ public function initialize(ExtensionManager $extensionManager): void
+                        ->scalarNode('path')->end()
+                        ->scalarNode('class')->end()
+                        ->scalarNode('env')->end()
+                        ->booleanNode('debug')->end()
```
default to `true` if omitted

adiq replied:
This default is defined in `SYMFONY_DEFAULTS` / `SYMFONY_4_DEFAULTS` -> `DEFAULT_DEBUG_MODE`. We can't set a default on input loading because we are losing information if given value has been entered by the user or is a default one (and we need that information in `autoconfigure` method).

adiq commented:
@pamil any updates? I would prefer not to use forks 😬

arnolanglade commented on README.md:
````diff
@@ -26,30 +26,40 @@ ensures that application behaviour will not be affected by stateful services.
 ```
 
-Symfony 4 does not have bootstrap file anymore and the environment is configured in the .env file.
+Symfony 4 is detected, based on the existence of `src/Kernel.php` file, so if You did not migrate to new Symfony structure yet; you need to set those values yourself.
````
You => you

arnolanglade commented on src/ServiceContainer/SymfonyExtension.php:
```diff
@@ -185,9 +233,9 @@ private function loadKernelContainer(ContainerBuilder $container): void
-            new Reference(self::KERNEL_ID),
-            'getContainer',
-        ]);
+                                             new Reference(self::KERNEL_ID),
```
you should not indent the code like that

arnolanglade commented on src/ServiceContainer/SymfonyExtension.php:
```diff
@@ -151,6 +168,37 @@ public function process(ContainerBuilder $container): void
+
+        $defaults = self::SYMFONY_DEFAULTS;
+
+        $symfony4KernelPath = sprintf('%s/%s', $container->getParameter('paths.base'), self::SYMFONY_4_DEFAULTS['kernel']['path']);
```
you should avoid this kind of abbreviation 4 => for

arnolanglade replied:
This is not true, you can put the symfony kernel where you want. The only thing that allow to know the version of symfony is a new parameter of the bootstrap parameter because it does not exist in SF4

arnolanglade replied:
missing return type hint :array

adiq replied:
@arnolanglade I have added Symfony 4 configuration defaults when `bootstrap: ~` is set and provided additional information in README as this is indeed an improvement and better DX.

Still, I want to also keep kernel file-based detection to provide zero-configuration setup for those using new default Symfony 4 structure.

adiq commented:
Any updates?

alanpoulain commented:
This PR would be very useful if merged.

lchrusciel commented:
Hello 👋 

@pamil is on vacation 🌴, and won't be available on GitHub until 10th od December. I'm not sure if I will find a time to check and review it before that date. Hopefully, we may be able to work on it during SymfonyCon or something.

adiq commented:
I heard that SymfonyCon was tremendous and the hack day is coming up shortly. 

Hope you had a good time there. If you anyone manage somehow to look at this PR, I would be delighted 😉

lchrusciel commented on src/ServiceContainer/SymfonyExtension.php:
```diff
@@ -143,6 +161,39 @@ public function process(ContainerBuilder $container): void
     }
 
+    /**
+     * @param ContainerBuilder $container
```
Let's remove these docblocks

lchrusciel commented on src/ServiceContainer/SymfonyExtension.php:
```diff
@@ -212,9 +260,9 @@ private function loadSharedKernelContainer(ContainerBuilder $container): void
-            new Reference(self::SHARED_KERNEL_ID),
-            'getContainer',
-        ]);
+                                             new Reference(self::SHARED_KERNEL_ID),
```
I bet something went wrong with formatting

lchrusciel commented:
Adrian, we will need a little rebase here ;)

adiq commented:
@lchrusciel I have updated this PR. Rebased from master, removed obsolete docblocks and fixed some coding style based on ECS. Tests are green, hope everything is fine now 🕺

pamil commented:
Thanks, Adrian! :tada:

pamil commented:
Sorry for the delay, I hope to get back to maintaining FriendsOfBehat now :)
